controllers/core/openstackclient_controller.go

```diff
@@ -0,0 +1,288 @@
+
+// OpenStackClientReconciler reconciles a OpenStackClient object
+type OpenStackClientReconciler struct {
+	client.Client
+	Scheme  *runtime.Scheme
+	Kclient kubernetes.Interface
+	Log     logr.Logger
+}
+
+//+kubebuilder:rbac:groups=core.openstack.org,resources=openstackclients,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core.openstack.org,resources=openstackclients/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=core.openstack.org,resources=openstackclients/finalizers,verbs=update
+//+kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;create;update;patch;delete;
+//+kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;list;watch;create;update;patch;delete;
+//+kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch;
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.12.2/pkg/reconcile
+func (r *OpenStackClientReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = r.Log.WithValues("openstackclient", req.NamespacedName)
+
+	instance := &corev1beta1.OpenStackClient{}
+	err := r.Client.Get(context.TODO(), req.NamespacedName, instance)
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			r.Log.Info("OpenStackClient CR not found", "Name", instance.Name, "Namespace", instance.Namespace)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	r.Log.Info("OpenStackClient CR values", "Name", instance.Name, "Namespace", instance.Namespace, "Secret", instance.Spec.OpenStackConfigSecret, "Image", instance.Spec.ContainerImage)
+
+	instance.Status.Conditions = condition.Conditions{}
+	cl := condition.CreateList(
+		condition.UnknownCondition(
+			corev1beta1.OpenStackClientReadyCondition,
+			condition.InitReason,
+			corev1beta1.OpenStackClientReadyInitMessage,
+		),
+	)
+	instance.Status.Conditions.Init(&cl)
+
+	h, err := helper.NewHelper(
+		instance,
+		r.Client,
+		r.Kclient,
+		r.Scheme,
+		r.Log,
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// Always patch the instance status when exiting this function so we can persist any changes.
+	defer func() {
+		// update the overall status condition if service is ready
+		if instance.Status.Conditions.IsTrue(corev1beta1.OpenStackClientReadyCondition) {
+			instance.Status.Conditions.MarkTrue(condition.ReadyCondition, condition.ReadyMessage)
+		}
+
+		if err := h.SetAfter(instance); err != nil {
+			util.LogErrorForObject(h, err, "Set after and calc patch/diff", instance)
+		}
+
+		if changed := h.GetChanges()["status"]; changed {
+			patch := client.MergeFrom(h.GetBeforeObject())
+
+			if err := r.Status().Patch(ctx, instance, patch); err != nil && !k8s_errors.IsNotFound(err) {
+				util.LogErrorForObject(h, err, "Update status", instance)
+			}
+		}
```

Comment:
use helper.PatchInstance() ?

Reply:
sure, that wasn't landed when I originally implemented last year but I can do it now